src/main/java/com/libertymutualgroup/herman/aws/ecs/PropertyHandler.java

```diff
@@ -15,16 +15,97 @@
  */
 package com.libertymutualgroup.herman.aws.ecs;
 
+import com.libertymutualgroup.herman.aws.AwsExecException;
+import com.libertymutualgroup.herman.logging.HermanLogger;
+import com.libertymutualgroup.herman.util.FileUtil;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashSet;
 import java.util.Properties;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class PropertyHandler {
+    private static final Pattern PROPERTY_PATTERN = Pattern.compile("\\$\\{([a-zA-Z0-9\\.\\_\\-]+)\\}");
+    private HermanLogger hermanLogger;
+    private String environmentName;
+    private String rootDirectory;
+    Properties props = new Properties();
+
+    public PropertyHandler(HermanLogger hermanLogger, String environmentName, String rootDirectory) {
+        this.hermanLogger = hermanLogger;
+        this.environmentName = environmentName;
+        this.rootDirectory = rootDirectory;
+    }
+
+    public void addProperty(String key, String value) {
+        props.put(key, value);
+    }
+
+    public String mapInProperties(String template) {
+        importPropFiles();
+
+        Set<String> propertiesToMatch = getPropertiesToMatch(template);
+        String result = template;
+        for (String prop : propertiesToMatch) {
+            String token = prop.replace("${", "");
+            token = token.replace("}", "");
+            Pattern groupPattern = Pattern.compile("\\$\\{(" + token + ")\\}");
+            Matcher matcher = groupPattern.matcher(result);
+            String value = lookupVariable(token);
+            if (value == null) {
+                throw new AwsExecException("Missing property set for " + token);
+            }
+            result = matcher.replaceAll(value);
+        }
+
+        if(getPropertiesToMatch(result).isEmpty()){
+            return result;
+        }
+
+        return mapInProperties(result);
```

Comment:
This is the key line - recur to continue mapping as long as the pattern is matched. The rest of the changes were to simplify and make it testable.